website/integrations: add opencloud integration (#22497)
goauthentik/authentik: website/integrations/chat-communication-collaboration/opencloud/index.mdx

---
title: Integrate with OpenCloud
sidebar_label: OpenCloud
support_level: community
---

## What is OpenCloud?

> OpenCloud is an open-source content collaboration platform for storing, syncing, and sharing files, built on the Infinite Scale (oCIS) architecture.
>
> -- https://opencloud.eu

## Preparation

The following placeholders are used in this guide:

- `opencloud.company` is the FQDN of the OpenCloud installation.
- `authentik.company` is the FQDN of the authentik installation.

This guide covers integrating authentik with the [`opencloud-compose`](https://github.com/opencloud-eu/opencloud-compose) Docker deployment. OpenCloud only supports authentication via OpenID Connect (OIDC).

Choose your setup below. The **Web only** tab logs in through the browser. The **Web, desktop & mobile** tab also enables the native sync clients, which each use a distinct client ID and require some extra issuer configuration.

> import TabItem from "@theme/TabItem";

Reviewer: lets move imports to top of file

Author: moved them up

    - **Application**: provide a name and note the **slug**.
    - **Choose a Provider type**: select **OAuth2/OpenID Connect**.
    - **Configure the Provider**:
        - **Client type**: `Public`
        - **Client ID**: `web`
        - **Redirect URIs**:
            - Strict: `https://opencloud.company/oidc-callback.html`
            - Strict: `https://opencloud.company/oidc-silent-redirect.html`
            - Strict: `https://opencloud.company/`
        - **Signing Key**: select any available key.
        - **Scopes**: `openid`, `profile`, `email`.
3. Click **Submit**.

## OpenCloud configuration

In the `opencloud-compose` project, enable the external IdP overlay in `COMPOSE_FILE`. This replaces OpenCloud's built-in IdP, so login goes through authentik only.

```bash
COMPOSE_FILE=docker-compose.yml:idm/external-idp.yml:custom/authentik-roles.yml
```

Set the OIDC values in `.env`:

```bash
OC_DOMAIN=opencloud.company
IDP_DOMAIN=authentik.company
IDP_ISSUER_URL=https://authentik.company/application/o/<application_slug>/
OC_OIDC_CLIENT_ID=web
OC_OIDC_CLIENT_SCOPES=openid profile email
WEBFINGER_WEB_OIDC_CLIENT_ID=web
WEBFINGER_WEB_OIDC_CLIENT_SCOPES=openid profile email
```

> :::info

Reviewer: no need to put this. its our doc and it should be considered authoritive, so no need to justify the values

Author: removed that for ya

`WEBFINGER_WEB_OIDC_CLIENT_ID` is required as the Web UI reads its client ID from WebFinger, and login will not start if it is empty.
:::

Create `custom/authentik-roles.yml` to assign every user the default role:

```yaml
---
services:
    opencloud:
        environment:
            PROXY_ROLE_ASSIGNMENT_DRIVER: "default"
            GRAPH_ASSIGN_DEFAULT_USER_ROLE: "true"
```

> Then reboot your docker containers.

Reviewer: Let's capitalize docker

Author: completed

import TabItem from "@theme/TabItem";
import Tabs from "@theme/Tabs";

<Tabs
  defaultValue="web"
  values={[
    { label: "Web only", value: "web" },
    { label: "Web, desktop & mobile", value: "all" },
  ]}
>
  <TabItem value="web">

## authentik configuration

1. Log in to authentik as an administrator and open the Admin interface.
2. Navigate to **Applications** > **Applications** and click **New Application**.
    - **Application**: provide a name and note the **slug**.
    - **Choose a Provider type**: select **OAuth2/OpenID Connect**.
    - **Configure the Provider**:
        - **Client type**: `Public`
        - **Client ID**: `web`
        - **Redirect URIs**:
            - Strict: `https://opencloud.company/oidc-callback.html`
            - Strict: `https://opencloud.company/oidc-silent-redirect.html`
            - Strict: `https://opencloud.company/`
        - **Signing Key**: select any available key.
        - **Scopes**: `openid`, `profile`, `email`.
3. Click **Submit**.

## OpenCloud configuration

> In the `opencloud-compose` project, enable the external IdP overlay in `COMPOSE_FILE`. This replaces OpenCloud's built-in IdP, so login goes through authentik only.

Reviewer: and that's by doing what's in L58. seems unclear

```bash
COMPOSE_FILE=docker-compose.yml:idm/external-idp.yml:custom/authentik-roles.yml
```

Set the OIDC values in `.env`:

```bash
OC_DOMAIN=opencloud.company
IDP_DOMAIN=authentik.company
IDP_ISSUER_URL=https://authentik.company/application/o/<application_slug>/
OC_OIDC_CLIENT_ID=web
OC_OIDC_CLIENT_SCOPES=openid profile email
WEBFINGER_WEB_OIDC_CLIENT_ID=web
WEBFINGER_WEB_OIDC_CLIENT_SCOPES=openid profile email
```

:::info

> `WEBFINGER_WEB_OIDC_CLIENT_ID` is required — the Web UI reads its client ID from WebFinger, and login will not start if it is empty.

Reviewer: ```suggestion
`WEBFINGER_WEB_OIDC_CLIENT_ID` is required as the Web UI reads its client ID from the WebFinger, and login will not start if it is empty.
```

Author: "the Webfinger" is not proper english, that's like telling someone "read the client id from the authentik."

Author: I replaced the emdash with as

:::

> Create `custom/authentik-roles.yml` to assign every user the default role:

Reviewer: if ur just adding env vars why not just add them to the env file instead

Author: This is how they specify doing it in their docs so this is what I did

```yaml
---
services:
    opencloud:
        environment:
            PROXY_ROLE_ASSIGNMENT_DRIVER: "default"
            GRAPH_ASSIGN_DEFAULT_USER_ROLE: "true"
```

> Recreate the stack:

Reviewer: usually we don't include these commands

Author: I removed them and said to just reboot your docker containers

```bash
docker compose up -d
```

## Configuration verification

> Open `https://opencloud.company` in a new browser window. You are redirected to authentik to log in, and after authenticating you are returned to OpenCloud.

Reviewer 2: Do they need to click anything to start the login process?

Author: Nah, it's an automatic redirect once it's configured. No option to log in with regular credentials setting this up

> Open `https://opencloud.company` in a new browser window. You are redirected to authentik to log in, and after authenticating you are returned to OpenCloud.

Reviewer: just say open opencloud no need to add domain, especially if it's a placeholder

Author: I'd prefer to leave it I think it's fine either direction but this is a bit more direct

  </TabItem>
  <TabItem value="all">

> OpenCloud's web, desktop, Android, and iOS clients each use a distinct client ID but must validate tokens against **one** issuer. authentik gives every application its own issuer by default, so this setup uses **GLOBAL issuer mode** (all providers share `https://authentik.company/`) plus a reverse proxy that serves OIDC discovery at that shared issuer.

Reviewer: ```suggestion
OpenCloud's web, desktop, Android, and iOS clients each use a distinct client ID, but must validate tokens against one issuer. authentik gives every application its own issuer by default, so this setup uses **GLOBAL issuer mode** (all providers share `https://authentik.company/`) plus a reverse proxy that serves OIDC discovery at that shared issuer.
```

is global issuer mode an authentik term or option? i don't think so. Everything after "so this setup" feels unclear

Author: So it is an authentik term inside of the code. The problem is that we don't seem to have any documentation on this field anywhere.

So under oauth2 provider -> advanced protocol settings -> issuer mode, the UI says "Same identifier is used for all providers", but it is saved as issuer mode "GLOBAL" on the provider. There is no documentation anywhere in our docs that highlights this option, and it is awkward to tell people that this setup uses the "Same identifier is used for all providers" issuer mode. So we don't really have a great option here besides call this global for now and then probably update our docs and UI later. Even the form field name in the UI is iffy to me imo

> ## 1. Serve discovery at the shared issuer

Reviewer 2: ```suggestion
## Reverse proxy configuration
```

> With GLOBAL issuer mode, tokens carry `iss = https://authentik.company/`, but authentik only serves discovery at `https://authentik.company/application/o/<application_slug>/.well-known/openid-configuration`. Put a reverse proxy in front of authentik that rewrites the root discovery path to one provider's discovery. Any reverse proxy works; for example, with Caddy:

Reviewer 2: ```suggestion
With GLOBAL issuer mode enabled, tokens use an issuer of `iss = https://authentik.company/`, but authentik only exposes OpenID Connect discovery at `https://authentik.company/application/o/<application_slug>/.well-known/openid-configuration`. To reconcile this, place a reverse proxy in front of authentik that maps the root discovery URL to a specific provider’s discovery endpoint. Any reverse proxy can handle this; for example, with Caddy:
```

```caddy

> # Forward authentik.company to this; it in turn forwards to authentik.

Reviewer: let's not include provider specific configuration for reverse proxies

Author: I think this makes it clear how the redirect should be set up so I want to keep this. It's a very small example and helps keeps things clear for the reader

Reviewer 2: Yeah I don't mind having specific instructions as an example. Plus if others want to contribute examples for nginx etc we can add them as tabs.

:8081 {
	@discovery path /.well-known/openid-configuration
	rewrite @discovery /application/o/<application_slug>/.well-known/openid-configuration
	reverse_proxy authentik-upstream:9000 {
		# keep authentik building https URLs if it is behind TLS termination
		header_up X-Forwarded-Proto https
	}
}
```

> ## 2. authentik configuration

Reviewer 2: ```suggestion
## authentik configuration
```

Repeat these steps for **each** of the four clients (Web, Desktop, Android, and iOS), using the per-client values from the table below.

1. Log in to authentik as an administrator and open the Admin interface.
2. Navigate to **Applications** > **Applications** and click **New Application**.
    - **Application**: provide a name and note the **slug**.
    - **Choose a Provider type**: select **OAuth2/OpenID Connect**.
    - **Configure the Provider**:
        - **Client type**: `Public`

>         - **Client ID**: the client's value from the table below.

Reviewer: table should be above the authentik block

Author: table moved

        - **Redirect URIs**: the client's value from the table below.

>         - **Signing Key**: select the **same** key for all four providers (the shared issuer exposes a single `jwks_uri`, so all clients' tokens must be signed by one key).

Reviewer: no need to justify and mabye just say Select the same available signing key for all platforms

Author: Sorry, I don't understand why we wouldn't, we are setting up a proxy to share one provider's well-known in a location to represent all 4 of these providers, it's important to note. It's the whole reason I had to make this guide in the first place

Reviewer 2: We don't usually give extra information in these steps. But in this case it's an unusual guide because it requires multiple providers. I'm fine with the justification.

>         - **Scopes**: `openid`, `profile`, `email`, and `offline_access` (required for the desktop and mobile sync clients to receive a refresh token).

Reviewer: no need
to justify scopes

Author: Sorry, why would there not be, I don't understand. I could not get this integration to work without offline_access and I don't understand how you would be able to skip this step

Reviewer 2: ```suggestion
        - Under **Advanced protocol settings**, add `offline_access` to **Selected scopes**.
```

@PeshekDotDev I think Dominic is talking about the "(required for the desktop and mobile sync clients to receive a refresh token)." 

We usually don't justify why a scope is being added.

>         - **Issuer mode** (under **Advanced protocol settings**): `Same identifier is used for all providers`.

Reviewer: Add section in previous line

- Under **Advanced proto settings**:

Author: I moved that up

3. Click **Submit**.

The per-client **Client ID** and **Redirect URIs**:

| Client  | Client ID          | Redirect URIs                                                                               |
| ------- | ------------------ | ------------------------------------------------------------------------------------------- |
| Web     | `web`              | Strict: `https://opencloud.company/oidc-callback.html`, `…/oidc-silent-redirect.html`, `…/` |
| Desktop | `OpenCloudDesktop` | Regex: `http://127.0.0.1(:[0-9]+)?(/.*)?` and `http://localhost(:[0-9]+)?(/.*)?`            |
| Android | `OpenCloudAndroid` | Strict: `oc://android.opencloud.eu`                                                         |
| iOS     | `OpenCloudIOS`     | Strict: `oc://ios.opencloud.eu`                                                             |

> ## 3. OpenCloud configuration

Reviewer 2: ```suggestion
## OpenCloud configuration
```

> In the `opencloud-compose` project, enable the external IdP overlay in `COMPOSE_FILE`. This replaces OpenCloud's built-in IdP, so login goes through authentik only.

Reviewer: ehhhh we repeat word for work what's in the web only section. we should restructure this doc as to not repeat ourselves

Author: There isn't really a good way to organize this to make that happen. The .env file for the web setup requires an application slug, the other .env file doesn't and has different urls.

The only thing that would kind of work is telling people to universally start with the authentik yml, then telling people to do some of opencloud, then go back to authentik, then go back to the .env for the web version, and I hate it when we do that in guides

```bash
COMPOSE_FILE=docker-compose.yml:idm/external-idp.yml:custom/authentik-roles.yml
```

Set the OIDC values in `.env`. `OC_OIDC_ISSUER` points at the shared (root) issuer, and each client uses its own WebFinger client ID:

```bash
OC_DOMAIN=opencloud.company
IDP_DOMAIN=authentik.company
IDP_ISSUER_URL=https://authentik.company/
OC_OIDC_CLIENT_ID=web
OC_OIDC_CLIENT_SCOPES=openid profile email

WEBFINGER_WEB_OIDC_CLIENT_ID=web
WEBFINGER_WEB_OIDC_CLIENT_SCOPES=openid profile email
WEBFINGER_DESKTOP_OIDC_CLIENT_ID=OpenCloudDesktop
WEBFINGER_DESKTOP_OIDC_CLIENT_SCOPES=openid profile email offline_access
WEBFINGER_IOS_OIDC_CLIENT_ID=OpenCloudIOS
WEBFINGER_IOS_OIDC_CLIENT_SCOPES=openid profile email offline_access
WEBFINGER_ANDROID_OIDC_CLIENT_ID=OpenCloudAndroid
WEBFINGER_ANDROID_OIDC_CLIENT_SCOPES=openid profile email offline_access
```

Create `custom/authentik-roles.yml` to assign every user the default role:

```yaml
---
services:
    opencloud:
        environment:
            PROXY_ROLE_ASSIGNMENT_DRIVER: "default"
            GRAPH_ASSIGN_DEFAULT_USER_ROLE: "true"
```

Recreate the stack:

```bash
docker compose up -d
```

## Configuration verification

> Open `https://opencloud.company` in a browser, and add the account in the Desktop, iOS, and Android apps using the same server URL. Each client is redirected to authentik to log in and returned to the client afterwards.

Reviewer 2: The "add the account" instruction should be in the "OpenCloud configuration" section.

Author: It's not possible to add the account until you've configured the above options for logging in with authentik, so this step has to be at the end

  </TabItem>
</Tabs>

## Resources

> - [OpenCloud documentation](https://docs.opencloud.eu/)

Reviewer: can we get the specific doc urls

Author: yeah good call, added
